### PR TITLE
Improve cf conformance 5.6 grid mappings and projections

### PR DESCRIPTION
**Pull request by leilabbb**

### CODE REVIEW

**Comment by benjwadams** [comment]
@leilabbb, could you address the test failures?

File: compliance_checker/tests/test_cf.py [hunk 1]
```diff
@@ -1293,16 +1293,18 @@ def test_check_reduced_horizontal_grid(self):
...
-        dataset = self.load_dataset(STATIC_FILES["mapping"])
-        results = self.cf.check_grid_mapping(dataset)
-
-        assert len(results) == 6
-        assert len([r.value for r in results.values() if r.value[0] < r.value[1]]) == 0
-        expected_name = (
-            "§5.6 Horizontal Coordinate Reference Systems, Grid Mappings, Projections"
-        )
-        assert all(r.name == expected_name for r in results.values())
+    # def test_check_grid_mapping(self):
```
Comment (benjwadams) [inline compliance_checker/tests/test_cf.py:1296]: Please remove commented section

File: compliance_checker/cf/cf_1_7.py [hunk 2]
```diff
@@ -237,9 +237,12 @@ def check_cell_boundaries(self, ds):
         # Note that test does not check monotonicity
         ret_val = []
         reasoning = []
+
         for variable_name, boundary_variable_name in cfutil.get_cell_boundary_map(
             ds
         ).items():
+            
+            print(boundary_variable_name, boundary_variable_name.dtype)
```
Comment (benjwadams) [inline compliance_checker/cf/cf_1_7.py:245]: Please remove print statement.  If needed for debug purposes, please use an appropriate low logging level such as `DEBUG` to suppress from normal output in the compliance checker.

File: compliance_checker/cf/cf_1_7.py [hunk 3]
```diff
@@ -501,6 +504,7 @@ def _check_gmattr_existence_condition_ell_pmerid_hdatum(self, var):
                     "reference_ellipsoid_name",
                     "prime_meridian_name",
                     "horizontal_datum_name",
+                    "geographic_crs_name",
```
Comment (benjwadams) [inline compliance_checker/cf/cf_1_7.py:507]: Add test coverage for added functionality.

**Review by benjwadams — CHANGES_REQUESTED** [review]

File: compliance_checker/cf/cf_1_7.py [hunk 4]
```diff
@@ -749,23 +844,39 @@ def check_grid_mapping(self, ds):
...
+                        test_ctx.out_of += 1                                   
+            test_ctx.score += 1
+ 
+            # existence_conditions           
             exist_cond_1 = (
                 self._check_gmattr_existence_condition_geoid_name_geoptl_datum_name(var)
             )
-            test_ctx.assert_true(exist_cond_1[0], exist_cond_1[1])
+            if exist_cond_1[0] == False:
+                test_ctx.messages.append("Both geoid_name and geopotential_datum_name "
```
Comment (benjwadams) [inline compliance_checker/cf/cf_1_7.py:855]: I think we already have this check defined here: https://github.com/ioos/compliance-checker/blob/master/compliance_checker/cf/cf_1_7.py#L568-L586

**Review by benjwadams — CHANGES_REQUESTED** [review]

File: compliance_checker/cf/cf_1_7.py (hunk 4 above)
Comment (leilabbb) [reply compliance_checker/cf/cf_1_7.py:855]: removed the IF conditional statement and used the tests' results from  _check_gmattr_existence_condition_ell_pmerid_hdatum and _check_gmattr_existence_condition_geoid_name_geoptl_datum_name

File: compliance_checker/cf/cf_1_7.py (hunk 2 above)
Comment (leilabbb) [reply compliance_checker/cf/cf_1_7.py:245]: removed "print( )", it was used for debugging.

File: compliance_checker/cf/cf_1_7.py (hunk 3 above)
Comment (leilabbb) [reply compliance_checker/cf/cf_1_7.py:507]: in progress ...

**Comment by benjwadams** [comment]
Can you check the test failures on this PR?

File: compliance_checker/cf/cf_1_7.py [hunk 5]
```diff
@@ -724,17 +728,105 @@ def _evaluate_towgs84(self, val):
...
+            # Recommendation: The grid mapping variables should have 0 dimensions.
+            if var.dimensions:
+               test_ctx.messages.append("The grid mapping variable {} should have 0 dimension".format(var.name))
+               test_ctx.out_of += 1
+            
+            test_ctx.score += 1
+
+            # [5/9] The grid mapping variables must have the grid_mapping_name attribute. 
+            # The legal values for the grid_mapping_name attribute are contained in Appendix F.
+            if "grid_mapping_name" not in var.ncattrs():
```
Comment (benjwadams) [inline compliance_checker/cf/cf_1_7.py:793]: `if not hasattr(var, "grid_mapping_name")` would be more Pythonic.

File: compliance_checker/cf/cf_1_7.py [hunk 6]
```diff
@@ -724,17 +728,105 @@ def _evaluate_towgs84(self, val):
...
+                if var.grid_mapping_name not in grid_mapping_dict17:
+                   test_ctx.messages.append("The legal values for the grid_mapping_name attribute "
+                    "{} is not contained in Appendix F".format(var.grid_mapping_name))             
+                   test_ctx.out_of += 1
+                
+            test_ctx.score += 1
+
+            # [6/9] The data types of the attributes of the 
+            # grid mapping variable must be specified in Table 1 of Appendix F.
+            type_map = {"S": "str" , "N": "float64"}
```
Comment (benjwadams) [inline compliance_checker/cf/cf_1_7.py:807]: The typecheck for `N` is likely too strong.  `N` simply means that the value has a numeric type, not just floating point.

File: compliance_checker/cf/cf_1_7.py [hunk 7]
```diff
@@ -749,33 +841,29 @@ def check_grid_mapping(self, ds):
                                 str(crs_error)
                             )
                         )
-                    else:
-                        test_ctx.score += 1
-                    test_ctx.out_of += 1
-
+                        test_ctx.out_of += 1
```
Comment (benjwadams) [inline compliance_checker/cf/cf_1_7.py:844]: Same comment about `hasattr` mentioned above.

File: compliance_checker/cf/cf_1_7.py [hunk 8]
```diff
@@ -749,33 +841,29 @@ def check_grid_mapping(self, ds):
...
-                test_ctx.out_of += 1
-                test_ctx.messages.append(
-                    "Cannot have both 'geoid_name' and "
-                    "'geopotential_datum_name' attributes in "
-                    "grid mapping variable '{}'".format(var.name)
-                )
-            elif len_vdatum_name_attrs == 1:
+            vert_datum_attrs = possible_vert_datum_attrs.intersection(var.ncattrs())                             
+            
+            if  exist_cond_1[0] == True and vert_datum_attrs:
```
Comment (benjwadams) [inline compliance_checker/cf/cf_1_7.py:866]: Just use `if exist_cond_1[0] and vert_datum_attrs` if `exist_cond[0]` is expected to be boolean.

**Review by benjwadams — CHANGES_REQUESTED** [review]

**Review by benjwadams — CHANGES_REQUESTED** [review]
Please also fix and/or change unit tests as necessary.

**Review by benjwadams — CHANGES_REQUESTED** [review]
EDIT: already mentioned

File: compliance_checker/cf/cf_1_7.py (hunk 5 above)
Comment (leilabbb) [reply compliance_checker/cf/cf_1_7.py:822]: agreed... changed

File: compliance_checker/cf/cf_1_7.py (hunk 6 above)
Comment (leilabbb) [reply compliance_checker/cf/cf_1_7.py:836]: agreed.... changed to:
type_map = {"S": "str", "N": ["float64", "int"]}

File: compliance_checker/cf/cf_1_7.py (hunk 7 above)
Comment (leilabbb) [reply compliance_checker/cf/cf_1_7.py:870]: agreed ... changed

**Review by leilabbb — COMMENTED** [review]
all comments were resolved except one that i am not sure about "Add test coverage for added functionality."

**Pull request closed without merging**